### PR TITLE
Use flat file listing in zarr file browser

### PR DESCRIPTION
**Pull request by AlmightyYakob**
This is a simplification, and will as a result fix any zarrs with out of sync file listings (any other zarrs affected by the underlying issue to https://github.com/dandi/dandi-archive/issues/1378).

@jwodder This is a breaking change and will require a corresponding CLI update. If anything here seem infeasible / unreasonable, please let me know.

Notable breaking changes:

* The url is now `zarr/{zarr_id}/files`
* To filter by prefix, use the `prefix` query param
* To limit the number of files returned, use the `limit` query param (default 1000)
* To perform pagination, provide the `after` query param, setting it to the key of the last item in the previous listing
* To be redirected to file download (requires providing file name in `prefix`), set the `download` flag to `true`
* The asset download endpoint no longer redirects to the zarr file explorer view if the specified asset contains a zarr. Instead, it returns a 400, instructing the user to use the zarr file listing endpoint directly. I prefer this than to blindly redirecting to zarr file listing endpoint, as the "download" operation on an entire zarr isn't something we support.

Head requests are still supported, and will use the provided `prefix` query param to return a presigned head URL. If the prefix is not the entire path of a file, the HEAD request against the presigned url will return a 404.

### CODE REVIEW

**Comment by jwodder** [comment]
@AlmightyYakob What is the format of the `zarr/{zarr_id}/files` response?

> To perform pagination, provide the `after` query param, setting it to the key of the last item in the previous listing

Is there a reason that pagination doesn't work like other endpoints, with `"results"` and `"next"` keys in the response?

**Comment by AlmightyYakob** [comment]
> @AlmightyYakob What is the format of the `zarr/{zarr_id}/files` response?

My bad. The format is as such:
```json
[
  {
    "Key": "foo/bar",
    "LastModified": "2022-12-06T21:46:34.331000Z",
    "Etag": "f0c0b28da13bec676d3a348d6f11dabe",
    "Size": 100
  }
]
```

> Is there a reason that pagination doesn't work like other endpoints, with `"results"` and `"next"` keys in the response?

Yes, S3 doesn't provide any mechanism for limit/offset pagination when listing objects, only providing a `StartAfter` parameter, to indicate the last element of the previous "page". There's also no way to get the total count of objects in the bucket (or under a certain prefix, as we would need), so it would be impossible to easily provide the `count` field, which is the standard for our other pagination endpoints.

**Comment by danlamanna** [comment]
I think we should still provide a `next` link where we can, this way it's easier for consumers and (hopefully) prevents them from hand-constructing URLs.

File: dandiapi/zarr/views/__init__.py [hunk 1]
```diff
@@ -97,13 +94,11 @@ class Meta:
...
-    directories = serializers.ListField(child=serializers.URLField())
-    files = serializers.ListField(child=serializers.URLField())
-    checksums = serializers.DictField(
-        child=serializers.RegexField('^[0-9a-f]{32}(-[0-9]+--[0-9]+)?$')
-    )
-    checksum = serializers.RegexField('^[0-9a-f]{32}-[0-9]+--[0-9]+$')
+class ZarrExploreQuerySerializer(serializers.Serializer):
+    after = serializers.CharField(default='')
+    prefix = serializers.CharField(default='')
+    limit = serializers.IntegerField(default=1000)
```
Comment (mvandenburgh) [inline dandiapi/zarr/views/__init__.py:100]: What happens if the client provides an extremely large value or a negative value here? Do we need `max_value` and `min_value`?

**Comment by AlmightyYakob** [comment]
@jwodder The endpoint response has been updated to the following

```json
{
  "next": "http://localhost:8000/api/zarr/357a0699-4f08-4a3b-83a7-923353379e28/files/?after=subdir%2Fa.txt&download=false&limit=1&prefix=subdir",
  "results": [
    {
      "Key": "subdir/a.txt",
      "LastModified": "2022-12-07 17:46:51.962000+00:00",
      "ETag": "c560516266eb28040362e72ec278b4a2",
      "Size": 8
    }
  ]
}
```

**Comment by jwodder** [comment]
@AlmightyYakob Does the `DELETE /zarr/:id/files/` endpoint for deleting Zarr entries still work the same way?

**Review by mvandenburgh — APPROVED** [review]
LGTM from the perspective of the API.

**Comment by jwodder** [comment]
@AlmightyYakob There was one failure in the test against https://github.com/dandi/dandi-cli/pull/1175, regarding the test of `get_content_url()`.  Has the contents of the `contentUrl` metadata field changed for Zarr assets?

**Comment by jwodder** [comment]
@AlmightyYakob The integration tests with my PR are passing now.

**Comment by AlmightyYakob** [comment]
> @AlmightyYakob The integration tests with my PR are passing now.

@jwodder So there's no changes needed on this PR, in reference to your previous comment?

**Comment by jwodder** [comment]
@AlmightyYakob Right.

**Pull request merged**